### DISCUSSION TITLE
Unable to read values into a List<ReferenceType> property using ClassMap

### DISCUSSION THREAD

**Discussion by enheh** [JoshClose/CsvHelper#1275] [Q&A]
If I try to use a class map to read values into a `List<ReferenceType>` property,
with something like this:
```
public class CarFleet ()
{
    public List<Velocity> Cars { get; set; }
}

public class Velocity ()
{
    public float Speed { get; set; }
    public float Direction { get; set; }
}

internal class CarFleetMap : ClassMap<CarFleet>
{
    public CarFleetMap()
    {
        Map(m => m.Cars[0].Speed).Name("car1_speed");
        Map(m => m.Cars[1].Speed).Name("car2_speed");
        Map(m => m.Cars[0].Direction).Name("car1_direction");
        Map(m => m.Cars[1].Direction).Name("car2_direction");
    }
}
```
I get a runtime exception like this: "Property 'Single Speed' is not defined for type 'CarFleet'"

I've looked all through the test cases and can't seem to find anything analogous.  Is there some way to make this work?

**Comment by AltruCoder** [comment]
A slight change to `CarFleet` may get you close to what you want.
```csharp
public class CarFleet 
{
    public Velocity Car1 { get; set; }
    public Velocity Car2 { get; set; }
}
```
You can use `ReferenceHeaderPrefix` to register the prefix for car speed and direction. `PrepareHeaderForMatch` will allow you to ignore the case of your titles.  Lastly, you can leave out the `CarFleetMap` class and just use `AutoMap()`
```csharp
using (var stream = new MemoryStream())
using (var writer = new StreamWriter(stream))
using (var reader = new StreamReader(stream))
using (var csv = new CsvReader(reader))
{
    csv.Configuration.Delimiter = ",";
    writer.WriteLine("car1_speed,car1_direction,car2_speed,car2_direction");
    writer.WriteLine("45.2,56.3,35.2,64.1");
    writer.Flush();
    stream.Position = 0;

    csv.Configuration.ReferenceHeaderPrefix = (type, name) => $"{name}_";
    csv.Configuration.PrepareHeaderForMatch = (header, index) => header.ToLower();
    csv.Configuration.AutoMap<CarFleet>();
    var records = csv.GetRecords<CarFleet>();
}
```

**Comment by enheh** [comment]
Using `ReferenceHeaderPrefix` and `PrepareHeaderForMatch` is a nice trick, and I'll definitely keep that in mind!  The code I presented here is just a simplified example to highlight the issue I'm encountering, though, and the real data structures are much more complicated requiring me to still use ClassMap.

Nonetheless, I can modify to the real version of `CarFleet` in the way you suggest as a workaround.  It would definitely be preferable, however, to have some way of reading reference types into an array or list, not just for this case, but in general for other potentially more involved scenarios.  Is that not possible?

**Comment by AltruCoder** [comment]
I am not the expert on this.  I have been answering questions mainly to increase my own knowledge.  There was an issue very similar to yours, but Josh only showed it was possible to map a single `ReferenceType` to the List.  #837 
It's a little messy, but using the ideas from that issue, I think I have what you are looking for.
```csharp
public class CarFleet
{
    public int Id { get; set; }
    public string Name { get; set; }
    public List<Velocity> Cars { get; set; }
}

using (var stream = new MemoryStream())
using (var writer = new StreamWriter(stream))
using (var reader = new StreamReader(stream))
using (var csv = new CsvReader(reader))
{
    csv.Configuration.Delimiter = ",";
    writer.WriteLine("Id,Name,car1_speed,car1_direction,car2_speed,car2_direction");
    writer.WriteLine("1,name,45.2,56.3,35.2,64.1");
    writer.Flush();
    stream.Position = 0;

    var records = new List<CarFleet>();
    csv.Read();
    csv.ReadHeader();
    while (csv.Read())
    {
        var record = csv.GetRecord<CarFleet>();
        record.Cars = new List<Velocity>();
        Velocity car = null;

        for (var i = 2; i < csv.Context.HeaderRecord.Count(); i++)
        {
            if (i % 2 == 0)
            {
                car = new Velocity
                {
                    Speed = csv.GetField<float>(i)
                };
            } 
            else
            {
                car.Direction = csv.GetField<float>(i);
                record.Cars.Add(car);
            }
        }
        records.Add(record);
    }
}
```

**Comment by exiled78** [comment]
I admit I did not read the doco, but I did not realise the Property name casing had to match the headers.
csvReader.Configuration.PrepareHeaderForMatch = (header, index) => header.ToLower(); fixed my issue too.

**Comment by JoshClose** [comment]
@enheh Are there a set number of cars on each row, or does it vary?
